File: snail-race-bet/java/src/main/java/snail/race/kata/adapters/RaceResultProviderHttpAntiCorruptionLayer.java
package snail.race.kata.adapters;

import org.jetbrains.annotations.NotNull;
import snail.race.kata.domain.RaceResultProvider;

import java.util.Comparator;
import java.util.List;

public class RaceResultProviderHttpAntiCorruptionLayer {
    static RaceResultProvider.SnailRaces mapToDomain(RaceResultProviderHttprRecords.Races races) {
        return new RaceResultProvider.SnailRaces(races.races().stream()
                .map(RaceResultProviderHttpAntiCorruptionLayer::mapSingleRace).toList());
    }

    private static RaceResultProvider.SnailRace mapSingleRace(RaceResultProviderHttprRecords.Race r) {
        return new RaceResultProvider.SnailRace(
                r.raceId(),
                r.timestamp(),
                createPodium(r.snails()));
    }

    private static RaceResultProvider.Podium createPodium(List<RaceResultProviderHttprRecords.Snail> snails) {
        List<RaceResultProvider.Snail> snailsOnPodium = snails.stream()
                .sorted(Comparator.comparingDouble(RaceResultProviderHttprRecords.Snail::duration))
                .limit(3)
                .map(s -> new RaceResultProvider.Snail(s.number(), s.name()))
                .toList();

        return new RaceResultProvider.Podium(
                snailsOnPodium.get(0),
                snailsOnPodium.get(1),
                snailsOnPodium.get(2)
        );
    }

}

File: snail-race-bet/typescript/src/adapters/RaceResultProviderHttpAntiCorruptionLayer.ts
import {Podium, Snail, SnailRace, SnailRaces} from "../domain/RaceResultProvider";

export type ApiSnail = {
    number: number;
    name: string;
    duration: number;
};
export type ApiRace = {
    raceId: number;
    timestamp: number;
    snails: ApiSnail[];
};
export type ApiRaces = {
    races: ApiRace[];
};

export class RaceResultProviderHttpAntiCorruptionLayer {
    static mapToDomain(data: ApiRaces): SnailRaces {
        return new SnailRaces(data.races.map(this.mapRaceToSnailRace));

    }

    private static mapRaceToSnailRace(race: ApiRace) {
        return new SnailRace(
            race.raceId,
            race.timestamp,
            RaceResultProviderHttpAntiCorruptionLayer.createPodium(race));
    }

    private static createPodium(race: ApiRace) {
        const sortedSnails = race.snails
            .sort((a: ApiSnail, b: ApiSnail) => a.duration - b.duration)
            .map(this.mapToSnail);

        return new Podium(
            sortedSnails[0],
            sortedSnails[1],
            sortedSnails[2]);
    }

    private static mapToSnail(snail: ApiSnail) {
        return new Snail(snail.number, snail.name);
    }

}